src/librustc_mir/monomorphize/deduplicate_instances.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright 2014 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use rustc_data_structures::indexed_vec::IndexVec;
+use rustc::ty::{self, TyCtxt, Ty, TypeFoldable, Instance, ParamTy};
+use rustc::ty::fold::TypeFolder;
+use rustc::ty::subst::Kind;
+use rustc::middle::const_val::ConstVal;
+use rustc::mir::{Mir, Rvalue, Promoted, Location};
+use rustc::mir::visit::{Visitor, TyContext};
+
+/// Replace substs which arent used by the function with TyError,
+/// so that it doesnt end up in the binary multiple times
+pub(crate) fn collapse_interchangable_instances<'a, 'tcx>(
```

Comment:
I'm glad you wrote a comment! But I think we could and should ultimately say more. I'd love to show a (small) example Rust program that would benefit, and then explain (in terms of that example) what this query inputs and outputs exactly.

e.g., some program like

```rust
fn foo<T>() { } // here, T is clearly unused =)

fn main() {
    foo::<u32>();
    foo::<u64>(); 
}
```